### PR TITLE
Add fail fast reporting to test runner.

### PR DESCRIPTION
**Pull request by kaspth**
### Inline output

Any failures or errors will be reported inline during the run by default.
Skipped tests will be reported if run in verbose mode.

Any result is output with failure messages and a rerun snippet for that test. I've tried to keep it minimal and relied on Minitest's summary reporter for details:

```
# Running:

.....S..........................................F

This failed

bin/rails test test/models/bunny_test.rb:14

.................................E

ArgumentError: Wups! Bet you didn't expect this!
    test/models/bunny_test.rb:19:in `block in <class:BunnyTest>'

bin/rails test test/models/bunny_test.rb:18

....................

Finished in 0.069708s, 1477.6019 runs/s, 1448.9106 assertions/s.

  1) Failure:
BunnyTest#test_something_failing [/Users/kasperhansen/Documents/code/collection_caching_test/test/models/bunny_test.rb:15]:
This failed


  2) Error:
BunnyTest#test_error_me:
ArgumentError: Wups! Bet you didn't expect this!
    test/models/bunny_test.rb:19:in `block in <class:BunnyTest>'

103 runs, 101 assertions, 1 failures, 1 errors, 1 skips

You have skipped tests. Run with --verbose for details.
```

Note that Minitest already informs us, we can get info about skipped tests if we run with `--verbose`. Handy!
### Fail fast

Aborts the test run on the first failure. Suit that groove with the last config option:

```
minitest options:
    -h, --help                       Display this help.
    -s, --seed SEED                  Sets random seed. Also via env. Eg: SEED=n rake
    -v, --verbose                    Verbose. Show progress processing files.
    -n, --name PATTERN               Filter run on /regexp/ or string.

Known extensions: pride, rails
    -p, --pride                      Pride. Show your testing pride!

Usage: bin/rails test [options] [files or directories]
You can run a single test by appending a line number to a filename:

    bin/rails test test/models/user_test.rb:27

You can run multiple files and directories at the same time:

    bin/rails test test/controllers test/integration/login_test.rb

By default test failures and errors are reported inline during a run.

Rails options:
    -e, --environment ENV            Run tests in the ENV environment
    -b, --backtrace                  Show the complete backtrace
    -f, --fail-fast                  Abort test run on first failure
```

Here's what it looks like:

```
# Run with bin/rails test --fail-fast
# Running:

................................................S......E

ArgumentError: Wups! Bet you didn't expect this!
    test/models/bunny_test.rb:19:in `block in <class:BunnyTest>'

bin/rails test test/models/bunny_test.rb:18

....................................F

This failed

bin/rails test test/models/bunny_test.rb:14

Interrupted. Exiting...


Finished in 0.051427s, 1808.3872 runs/s, 1769.4972 assertions/s.

  1) Error:
BunnyTest#test_error_me:
ArgumentError: Wups! Bet you didn't expect this!
    test/models/bunny_test.rb:19:in `block in <class:BunnyTest>'


  2) Failure:
BunnyTest#test_something_failing [/Users/kasperhansen/Documents/code/collection_caching_test/test/models/bunny_test.rb:15]:
This failed

93 runs, 91 assertions, 1 failures, 1 errors, 1 skips
```

cc @senny @arthurnn @dhh

### CODE REVIEW

**Comment by kaspth** [comment]
Hm, it seems Travis is creating a build for a merge commit rather than the one I've written. This Travis build, https://travis-ci.org/rails/rails/builds/77903799, is running based on: https://github.com/rails/rails/commit/db989a2ed29e44e73cf4c8a75dfa2d20da68d9ae

Which says "Merge 485b662 into 79f44eb" where 485b662 is the commit here. That commit mismatch must be why the build status doesn't show up here!

**Comment by rafaelfranca** [comment]
@joshk maybe this is the reason?

**Comment by rafaelfranca** [comment]
I think it is already possible to do that today without having to implement it. If you send SIGINFO (control + T) on Mac OS, it prints the current test failures. Not sure we should implement the same thing again.

**Comment by kaspth** [comment]
Yes, it does show failures and errors but without any rerun commands:

```
........E........load: 2.17  cmd: ruby 8681 running 1.53u 0.42s
Current results:


  1) Failure:
BunnyTest#test_something_failing [/Users/kasperhansen/Documents/code/collection_caching_test/test/models/bunny_test.rb:16]:
This failed


  2) Error:
BunnyTest#test_error_me:
ArgumentError: Wups! Bet you didn't expect this!
    test/models/bunny_test.rb:20:in `block in <class:BunnyTest>'



Current: BunnyTest#test_this_thing_works_79 0.07s
......................

Finished in 10.386975s, 9.9163 runs/s, 9.7237 assertions/s.
```

**Comment by senny** [comment]
I didn't know about `Ctrl-T` :flushed:. @kaspth there's no way to extend that with the rerun snippets? Also, would be possible to enable fail-fast by default when using `Ctrl-T`?

**Comment by kaspth** [comment]
I believe Minitest does have a signal trap which lets you enrich a signal. I'll look into it.

I'd still like the option to not have to press `Ctrl + T` (which I believe is only an OS X shortcut) to see the failures fast.

**Comment by rafaelfranca** [comment]
Also fail-fast in RSpec means, stop the test execution on the first test failure so if we are going to use an option I'd recommend to not use the same name as RSpec

**Comment by kaspth** [comment]
Hm, right. We could go with `--defer-output` for disabling this and perhaps have a real fail fast that just raises `Interrupt` on the first failure.

**Comment by arthurnn** [comment]
> have a real fail fast that just raises Interrupt
:+1:

File: railties/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,31 @@
+*   Add fail fast output to `bin/rails test`
```
Comment (arthurnn) [inline railties/CHANGELOG.md]: this needs to change to `fail slow` i guess.

Comment (arthurnn) [inline railties/CHANGELOG.md]: NVM

**Comment by arthurnn** [comment]
Yeah, for sure I was confused here.
`--fail-fast` for me, should do the same as Rspec.. which means. exit right in the first fail/error.

The output, I guess we should not change much from what is default on Minitest.
which is.. only show fail/errors in the end, or if you pass a `--verbose` will show the fail/errors as they happen.

**Comment by kaspth** [comment]
Alright, I've added a fail fast mode which will abort the run on the first failure. I've taken a literal approach meaning failing doesn't trigger on errors (which might be silly, come to think of it).

**Comment by kaspth** [comment]
I don't know if we should do something to improve the output on Minitest `< 5.6.0`. Check the changelog where there's an example of a backtrace that's spewed in those versions.

**Comment by rafaelfranca** [comment]
> I don't know if we should do something to improve the output on Minitest < 5.6.0. Check the changelog where there's an example of a backtrace that's spewed in those versions.

I'd not even mention that in CHANGELOG. I always expect people to keep their dependecies up to date.

**Comment by rafaelfranca** [comment]
:shipit:

**Comment by kaspth** [comment]
Good suggestion, @rafaelfranca. Will amend the changelog and merge! :grin:

**Comment by senny** [comment]
@kaspth awesome! :yellow_heart: :tada:

**Comment by kaspth** [comment]
:heart:

**Comment by arthurnn** [comment]
this is awesome

**Comment by teoljungberg** [comment]
Hey, I don't want to be that guy - but I think I should get some recognition for previous work: https://github.com/teoljungberg/minitest-fail-fast

/cc @kaspth

**Comment by tenderlove** [comment]
@teoljungberg is this a derivation of your work? Or you just want people to know that you wrote one too?  Just trying to understand why, since we'd probably also have to mention RSpec, Minitest::Reporters, [this blog post](http://www.opinionatedprogrammer.com/2010/12/minitest-test-unit-immediate-backtrace-and-exit-on-failure/), etc.

**Comment by teoljungberg** [comment]
@tenderlove I think mentioning previous work in general is not a bad idea. Though, in the view of the other implementations out there - mine's the least important.

**Pull request merged**